### PR TITLE
feat: add OP_MERKLEPATHVERIFY opcode

### LINKED ISSUE #6: OP_MERKLEPATHVERIFY: Generalized Merkle path verification opcode
## Summary

Proposal for Merkle tree verification supporting both Taproot taptree script-path verification and Arkade metadata inclusion proofs. Both use BIP-341 tagged hashes with domain-specific tags.

This proposal shows two approaches:
1. **Composed from existing opcodes** — works today with `OP_SHA256INITIALIZE`/`UPDATE`/`FINALIZE`
2. **Dedicated `OP_MERKLEPATHVERIFY` opcode** — cleaner scripts, same semantics

Related: #2, [arkade-assets PR #16](https://github.com/ArkLabsHQ/arkade-assets/pull/16)

## Background

Both Taproot and Arkade metadata trees use the same BIP-341 tagged hash pattern with domain-specific tags:

| | Leaf tag | Branch tag |
|---|---|---|
| **Taproot** | `"TapLeaf"` | `"TapBranch"` |
| **Arkade metadata** | `"ArkadeAssetLeaf"` | `"ArkadeAssetBranch"` |

```
tagged_hash(tag, msg) = SHA256( SHA256(tag) || SHA256(tag) || msg )
```

Branches are lexicographically sorted (smaller hash first), eliminating direction bits from proofs.

---

## Approach 1: Composed from Existing Opcodes

Tagged hashes can be built from the existing incremental SHA256 opcodes. No new opcodes required.

### Tagged Hash Pattern

```
// tagged_hash(tag, msg) using existing opcodes
// Expects stack: [..., msg]

<tag>                                     // push tag bytes
OP_SHA256                                 // tag_hash = SHA256(tag)
OP_DUP                                   // duplicate for double-prefix
OP_SHA256INITIALIZE                       // ctx = SHA256_init(tag_hash)
OP_SWAP                                  // bring second tag_hash to top
OP_SHA256UPDATE                           // ctx = SHA256_update(ctx, tag_hash)
OP_SHA256FINALIZE                         // result = SHA256_finalize(ctx, msg)
```

### Sorted Branch Pattern

```
// branch_hash = tagged_hash(branch_tag, sorted(a, b))
// Expects stack: [..., a, b]

OP_2DUP OP_LESSTHAN
OP_IF
  OP_SWAP
OP_ENDIF
OP_CAT                                    // sorted_pair = min(a,b) || max(a,b)

<branch_tag> OP_SHA256
OP_DUP OP_SHA256INITIALIZE
OP_SWAP OP_SHA256UPDATE
OP_SHA256FINALIZE                         // branch_hash
```

### Full Proof Verification (2-leaf tree example)

Verifies a single metadata key-value pair is in a 2-leaf Arkade metadata tree:

```
// Inputs: <leaf_data> <sibling_hash> <expected_root>

// Step 1: leaf_hash = tagged_hash("ArkadeAssetLeaf", leaf_data)
OP_ROT
<"ArkadeAssetLeaf"> OP_SHA256
OP_DUP OP_SHA256INITIALIZE
OP_SWAP OP_SHA256UPDATE
OP_SHA256FINALIZE                         // stack: [root, sibling, leaf_hash]

// Step 2: branch = tagged_hash("ArkadeAssetBranch", sorted(leaf_hash, sibling))
OP_2DUP OP_LESSTHAN
OP_IF
  OP_SWAP
OP_ENDIF
OP_CAT
<"ArkadeAssetBranch"> OP_SHA256
OP_DUP OP_SHA256INITIALIZE
OP_SWAP OP_SHA256UPDATE
OP_SHA256FINALIZE                         // stack: [root, computed_root]

// Step 3: verify
OP_EQUALVERIFY
```

### Full Proof Verification (deeper trees)

For a proof with N siblings, repeat the sorted branch pattern N times. Each iteration pops the current hash and next sibling, producing the parent hash:

```
// Inputs: <leaf_data> <sib_0> <sib_1> ... <sib_N-1> <expected_root>

// Compute leaf_hash (tagged hash with leaf tag, same as above)
...

// For each sibling (repeat N times):
//   sorted branch hash (8-9 ops per level)
...

// Final verify
OP_EQUALVERIFY
```

A 3-sibling proof (8-leaf tree) requires ~30 opcodes total.

### Trade-offs

- **Pro**: Works today, no consensus changes needed
- **Con**: Verbose scripts (~8 ops per tree level + ~6 ops for leaf hash)
- **Con**: Tag string pushed repeatedly in script (can precompute `SHA256(tag)` as a constant to save ~1 op per use)

### Optimization: Precomputed Tag Hashes

Since tag strings are known at script creation time, `SHA256(tag)` can be precomputed and embedded as a 32-byte constant, saving one `OP_SHA256` per tagged hash:

```
// Precomputed: TAG_HASH = SHA256("ArkadeAssetLeaf")
<TAG_HASH>
OP_DUP OP_SHA256INITIALIZE
OP_SWAP OP_SHA256UPDATE
OP_SHA256FINALIZE                         // tagged_hash result (5 ops instead of 6)
```

---

## Approach 2: Dedicated `OP_MERKLEPATHVERIFY`

A single opcode that encapsulates leaf hashing, proof walking, and root verification.

### Stack Inputs

```
<leaf_tag> <branch_tag> <proof> <leaf_data> <expected_root> OP_MERKLEPATHVERIFY
```

| Input | Description |
|-------|-------------|
| `expected_root` | 32-byte expected Merkle root |
| `leaf_data` | Serialized leaf payload (caller encodes per their domain's format) |
| `proof` | Concatenated 32-byte sibling hashes (length must be multiple of 32) |
| `branch_tag` | ASCII tag for branch hashing (e.g. `"TapBranch"`, `"ArkadeAssetBranch"`) |
| `leaf_tag` | ASCII tag for leaf hashing (e.g. `"TapLeaf"`, `"ArkadeAssetLeaf"`) |

### Verification Algorithm

```python
def op_merklepathverify(leaf_tag, branch_tag, proof, leaf_data, expected_root):
    assert len(expected_root) == 32
    assert len(proof) % 32 == 0

    # Compute leaf hash
    leaf_hash = tagged_hash(leaf_tag, leaf_data)

    # Walk the proof path
    current = leaf_hash
    for i in range(0, len(proof), 32):
        sibling = proof[i:i+32]
        if current < sibling:
            current = tagged_hash(branch_tag, current || sibling)
        else:
            current = tagged_hash(branch_tag, sibling || current)

    assert current == expected_root
```

### Usage Examples

**Taproot script-path verification:**
```
<"TapLeaf"> <"TapBranch"> <proof> <leaf_data> <taptree_root> OP_MERKLEPATHVERIFY
```

**Arkade metadata verification:**
```
<"ArkadeAssetLeaf"> <"ArkadeAssetBranch"> <proof> <leaf_data> OP_INSPECTASSETGROUPMETADATAHASH OP_MERKLEPATHVERIFY
```

### Error Conditions

- `expected_root` is not 32 bytes → script failure
- `proof` length is not a multiple of 32 → script failure
- `leaf_tag` or `branch_tag` is empty → script failure
- Computed root ≠ expected root → script failure

### Trade-offs

- **Pro**: Clean, minimal scripts (1 opcode vs ~30)
- **Pro**: Less witness data (tag strings pushed once, not per level)
- **Pro**: Easier to audit script correctness
- **Con**: Requires new opcode (consensus change)
- **Con**: Less flexible than composing primitives

---

## Leaf Data Formats

The verification is agnostic to leaf encoding — it hashes whatever bytes it receives. Each domain defines its own serialization:

**Taproot (`"TapLeaf"`):**
```
leaf_data = leaf_version || compact_size(script_len) || script
```

**Arkade metadata (`"ArkadeAssetLeaf"`):**
```
leaf_data = leaf_version || varuint(key_len) || key || varuint(value_len) || value
```

## Tagged Hash Primitive

For reference, the tagged hash construction from BIP-341:

```
tagged_hash(tag, msg) = SHA256(SHA256(tag) || SHA256(tag) || msg)
```

Implementations SHOULD cache `SHA256(tag)` for repeated use.

### PR DESCRIPTION
**Pull request by Kukks**
## Summary

- Implement `OP_MERKLEPATHVERIFY` (opcode `0xf4`) -- a dedicated opcode for generalized Merkle path verification using BIP-341 tagged hashes
- Add comprehensive tests covering valid verification (2-leaf tree, 4-leaf tree, single-leaf/empty proof) and all error conditions (wrong root, bad proof length, empty tags, wrong root size)
- Update `TestOpcodeDisasm` to recognize the new opcode

### Stack inputs (top to bottom)

```
<leaf_tag> <branch_tag> <proof> <leaf_data> <expected_root> OP_MERKLEPATHVERIFY
```

The opcode computes a leaf hash via `tagged_hash(leaf_tag, leaf_data)`, walks the proof path by sorting and hashing pairs with the branch tag, and verifies the result matches the expected root. On success the stack is consumed; on failure the script aborts.

Closes #6

## Test plan

- [x] `TestMerklePathVerify/valid_2leaf_tree` -- verifies a 2-leaf Merkle tree
- [x] `TestMerklePathVerify/valid_4leaf_tree` -- verifies a 4-leaf Merkle tree with 2-sibling proof
- [x] `TestMerklePathVerify/valid_empty_proof_single_leaf` -- verifies empty proof where leaf hash equals root
- [x] `TestMerklePathVerify/invalid_wrong_expected_root` -- rejects mismatched root
- [x] `TestMerklePathVerify/invalid_proof_length_not_multiple_of_32` -- rejects proof with bad length
- [x] `TestMerklePathVerify/invalid_empty_leaf_tag` -- rejects empty leaf tag
- [x] `TestMerklePathVerify/invalid_empty_branch_tag` -- rejects empty branch tag
- [x] `TestMerklePathVerify/invalid_expected_root_not_32_bytes` -- rejects root that is not 32 bytes
- [x] `TestOpcodeDisasm` -- confirms opcode disassembles as `OP_MERKLEPATHVERIFY`
- [x] All existing tests continue to pass

### CODE REVIEW

File: pkg/arkade/opcode.go [hunk 1]
```diff
@@ -3118,3 +3118,94 @@ func opcodeSha256Finalize(op *opcode, _ []byte, vm *Engine) error {
...
+
+// taggedHash computes a BIP-341 tagged hash: SHA256(SHA256(tag) || SHA256(tag) || msg)
+func taggedHash(tag, msg []byte) []byte {
+	tagHash := sha256.Sum256(tag)
+	h := sha256.New()
+	h.Write(tagHash[:])
+	h.Write(tagHash[:])
+	h.Write(msg)
+	return h.Sum(nil)
+}
```
Comment (louisinger) [inline pkg/arkade/opcode.go]: `chainhash.TaggedHash`

**Review by louisinger — CHANGES_REQUESTED** [review]
update README
